### PR TITLE
[Dropbox] Fix fetching documents from team shared folders

### PR DESCRIPTION
**Pull request by moxarth-elastic**
## Closes https://github.com/elastic/connectors/issues/2704

This PR now fetches all the files present at the team shared folders. 
Earlier we pass folder_id for the header - `Dropbox-API-Path-Root`, which was causing this issue of missing root folder content. Now, we're passing `root_namespace_id` for the header which includes all team shared folders content.

#### Pre-Review Checklist
- [x] this PR does NOT contain credentials of any kind, such as API keys or username/passwords (double check `config.yml.example`)
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] Considered corresponding documentation changes
- [ ] Contributed any configuration settings changes to the configuration reference
- [ ] if you added or changed Rich Configurable Fields for a Native Connector, you made a corresponding PR in [Kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/native_connectors.ts)

## Release Note

Dropbox connector now fetches the files from the team shared folders.

### CODE REVIEW

File: connectors/sources/dropbox.py [hunk 1]
```diff
@@ -844,11 +854,41 @@ async def close(self):
...
+    async def set_user_info(self):
+        try:
             endpoint = EndpointName.AUTHENTICATED_ADMIN.value
-        await self.dropbox_client.ping(endpoint=endpoint)
-        self._logger.info("Successfully connected to Dropbox")
+            response = await self.dropbox_client.ping(endpoint=endpoint)
+        except ClientResponseError as exception:
+            if (
+                str(exception.request_info.url)
+                == "https://api.dropboxapi.com/2/team/token/get_authenticated_admin"
```
Comment (navarone-feekery) [inline connectors/sources/dropbox.py:879]: I feel like this URL should be a const

File: connectors/sources/dropbox.py [hunk 2]
```diff
@@ -1043,9 +1085,7 @@ async def get_account_details(self):
 
     async def get_permission_list(self, item_type, item, account_id):
         if item_type == FOLDER:
-            shared_folder_id = item.get("shared_folder_id") or item.get(
-                "parent_shared_folder_id"
-            )
```
Comment (navarone-feekery) [inline connectors/sources/dropbox.py:1048]: Why was this removed?

File: connectors/sources/dropbox.py [hunk 3]
```diff
@@ -1146,20 +1192,19 @@ async def add_document_to_list(self, func, account_id, folder_id, is_shared=Fals
...
-                func=self._fetch_shared_files,
-                account_id=account_id,
-                folder_id=folder_id,
-                is_shared=True,
-            ):
-                yield mapped_document
+        async for mapped_document in self.add_document_to_list(
+            func=self._fetch_files_folders,
+            account_id=account_id,
+            folder_id=self.dropbox_client.member_id,
```
Comment (navarone-feekery) [inline connectors/sources/dropbox.py:1198]: Why are we using the `self.dropbox_client.member_id` here instead of the original `folder_id`?

**Review by navarone-feekery — COMMENTED** [review]
I've left a few questions because I'm quite unfamiliar with dropbox API. It'd be a great help if you expand the PR description to explain why you've made many of these changes.

File: connectors/sources/dropbox.py (hunk 3 above)
Comment (moxarth-elastic) [reply connectors/sources/dropbox.py:1198]: Seems like these changes does not change anything, i tried to revert the changes and tested the connector; it also works properly. So, reverting this changes to reduce confusion :)

**Comment by moxarth-elastic** [comment]
> It'd be a great help if you expand the PR description to explain why you've made many of these changes.

I've mentioned the root cause of the issue in the description; and some of the changes are of code auto-formatting (make autoformat).

File: connectors/sources/dropbox.py [hunk 4]
```diff
@@ -844,11 +857,35 @@ async def close(self):
...
             endpoint = EndpointName.AUTHENTICATED_ADMIN.value
-        await self.dropbox_client.ping(endpoint=endpoint)
-        self._logger.info("Successfully connected to Dropbox")
+            await self.dropbox_client.ping(endpoint=endpoint)
+            self._logger.info("Successfully connected to Dropbox")
+        except ClientResponseError as exception:
+            if str(exception.request_info.url) == AUTHENTICATED_ADMIN_URL:
+                endpoint = EndpointName.PING.value
+                await self.dropbox_client.ping(endpoint=endpoint)
+                self._logger.info("Successfully connected to Dropbox")
```
Comment (navarone-feekery) [inline connectors/sources/dropbox.py:868]: Why do we attempt to ping a different endpoint if the first one fails?

Also, `endpoint` has already been set to `endpoint = EndpointName.AUTHENTICATED_ADMIN.value` on L861, but if there is a client response error the code checks `if str(exception.request_info.url) == AUTHENTICATED_ADMIN_URL:`. Can the value from `exception.request_info.url` ever _not_ be the authenticated admin URL used in the request?

File: tests/sources/test_dropbox.py [hunk 5]
```diff
@@ -1215,9 +1220,14 @@ async def test_advanced_rules_validation_with_invalid_repos(
         [JSONAsyncMock(MOCK_SEARCH_FILE_3, 200)],
     ),
 )
+@patch.object(
+    DropboxClient,
+    "ping",
+    return_value=JSONAsyncMock(MOCK_AUTHENTICATED_ADMIN, 200),
```
Comment (navarone-feekery) [inline tests/sources/test_dropbox.py:1226]: Can you add tests covering the following for the methods `ping` and `set_user_info`:

1. When the authenticated admin ping returns a `ClientResponseError`
2. When the authenticated admin ping returns a different error (e.g. `TimeoutError`)
3. When the authenticated admin ping returns a `ClientResponseError` and the `exception.request_info.url` was an unexpected value

File: connectors/sources/dropbox.py [hunk 6]
```diff
@@ -844,11 +857,35 @@ async def close(self):
...
+    async def set_user_info(self):
+        try:
+            endpoint = EndpointName.AUTHENTICATED_ADMIN.value
+            response = await self.dropbox_client.ping(endpoint=endpoint)
+        except ClientResponseError as exception:
+            if str(exception.request_info.url) == AUTHENTICATED_ADMIN_URL:
+                endpoint = EndpointName.PING.value
+                response = await self.dropbox_client.ping(endpoint=endpoint)
+            else:
+                raise
```
Comment (navarone-feekery) [inline connectors/sources/dropbox.py:881]: Same as previous question.

Comment (navarone-feekery) [reply connectors/sources/dropbox.py:881]: Also, if the first request returns a `ClientResponseError`, `response` will become the result from the ping endpoint instead of the authenticated admin endpoint. Will that not cause issues when trying to extract values for `self.dropbox_client.member_id` and `self.dropbox_client.root_namespace_id`?

File: connectors/sources/dropbox.py [hunk 7]
```diff
@@ -844,11 +857,35 @@ async def close(self):
...
+        _json = await response.json()
+        self.dropbox_client.member_id = _json.get("admin_profile", {}).get(
+            "team_member_id"
+        )
+        self.dropbox_client.root_namespace_id = _json.get("root_info", {}).get(
+            "root_namespace_id"
+        ) or _json.get("admin_profile", {}).get("root_folder_id")
```
Comment (navarone-feekery) [inline connectors/sources/dropbox.py:888]: The following two dict calls can raise an raise an error:

1. `_json.get("admin_profile", {}).get("team_member_id")` if `{ admin_profile: None }`
2. `_json.get("admin_profile", {}).get("root_folder_id")` if `{ admin_profile: None }`
3. `_json.get("root_info", {}).get("root_namespace_id")` if `{ root_info: None }`

Can you null-safe these?

Also, if the json has `{ admin_profile: None }`, does that effectively mean Dropbox cannot sync documents? If so I think we need the following changes, otherwise the stack trace will not provide enough information to a user about what went wrong.

1. A detailed log explaining what is wrong with the response format
2. Raise a custom error (e.g. `DropboxResponseMalformed` error or something)

**Review by navarone-feekery — COMMENTED** [review]
Thanks for the extra summary on the PR description!
I've done another review now that I have more context. I can't test this against a dropbox instance myself, so can you also provide a log output example?

File: connectors/sources/dropbox.py (hunk 4 above)
Comment (moxarth-elastic) [reply connectors/sources/dropbox.py:868]: https://github.com/elastic/connectors/blob/76ef0c1618e1d67d92c1bd066ea820871fc30008/connectors/sources/dropbox.py#L846-L851

The previous ping method was checking if the DLS is enabled; we use `EndpointName.AUTHENTICATED_ADMIN.value` endpoint else `EndpointName.PING.value` endpoint. 

So, the problem arrives when the business account is used and DLS is not enabled. FYI, to fetch content from business account, the token should have team permission and so it will not work for the [PING](https://www.dropbox.com/developers/documentation/http/documentation#users-get_current_account) endpoint ; even though the connection can be established, the connector throws ClientRequestError for `EndpointName.PING.value` endpoint in that case. 

To fix this, we have two options:
1. Add an RCF for the Dropbox account type; either Basic or Business 
2. (current approach) We can check the `EndpointName.AUTHENTICATED_ADMIN.value` endpoint, if we get ClientResponseError then we move forward with the another endpoint - `PING`.

Comment (navarone-feekery) [reply connectors/sources/dropbox.py:868]: @moxarth-elastic thanks for the detailed explanation. Sorry, I didn't realise from the diff that this was how it already was, I thought the addition of two different ping endpoints was from this PR.

**Review by navarone-feekery — APPROVED** [review]
LGTM
